### ISSUE TITLE
DockerHub rate limit

### ISSUE THREAD

**Issue by anthr76** [traefik/traefik#8149]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

https://github.com/traefik/traefik/issues/7566

```
Failed to pull image "traefik:2.4.8": rpc error: code = Unknown desc = failed to pull and unpack image "docker.io/library/traefik:2.4.8": failed to copy: httpReaderSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/library/traefik/manifests/sha256:08d8a7759f5fffa2441488151cedcd4d556c1f124c097f929f469c1f7b82c16f: 429 Too Many Requests - Server message: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```

**Not** this. It has been almost a year since DockerHub rate limits have been imposed. This is a huge pain point to have your ingresscontroller be rate limited causing downtime. According to previous issues traefik was supposed to not be rate limited. This is not the case. Is it possible to push to another registry in CI? Having another registry helps incase DockerHub suffers downtime.

**Comment by mpl** [comment]
@anthr76 Hello. Having the image hosted officially anywhere else will take take some time (if ever), as we would have to study all the alternatives, and the related pros and cons (including hosting and egress costs, etc), before we can actually do it.
Our recommendation for now is to pull while being authenticated with dockerhub, as it should give you a more lenient rate limit.

**Comment by onedr0p** [comment]
@mpl I wouldn't call an additional 100 pulls per 6 hours more lenient while authenticated. Why can traefik use GHCR?

**Comment by mpl** [comment]
> @mpl I wouldn't call an additional 100 pulls per 6 hours more lenient while authenticated. Why can't traefik also use GHCR?

we can't just also use <alternative>, because any alternative that has the best rate-limiting for the users will de facto become the solution that everyone wants to use. Therefore it basically becomes our official image(s), and that is why we have to study them and choose the best one we can. Otherwise we'll basically end up in the same situation, just with another provider.

> Traefik maintainers can also apply for the Dockerhub Open source developer license which will remove the pull limits for Traefik.
> 
> https://www.docker.com/blog/expanded-support-for-open-source-software-projects/

AFAIK (but I could be wrong), we already did that but never got any response so far. But I'll ask again where we are on that.

**Comment by anthr76** [comment]
Hi @mpl It's seems this isn't something traefik or you are really interested in doing for one reason or not. Perhaps things are a bit hectic and you can't be bothered for a few google searches.

This is somewhat of a big deal. Authenticated dockerhub pulls are great until your within a corporate network and share one public IPv4. If you have clusters behind that you quickly reach a rate-limit authenticated or not. Like @onedr0p eluded the quickest and best way to fix this issue is to actually follow through with https://github.com/traefik/traefik/issues/7566 a locked issue from almost a year ago.

It seems like Traefik+Dockerhub doesn't have the capacity to make the " expanded support for open source software projects" work. That's okay. Considering the source code for Traefik lives on GitHub in my opinion it makes total sense to push to both Dockerhub and GHCR. Traefik has (a long time ago) outgrown traefik. It's deployed in OCI and non infrastructure. Limiting your image to only DockerHub would have made sense in 2017 but not now. It didn't even before rate limits were imposed considering your supported platforms. 

I can't find one thing on the internet for egress limits with GHCR. With about 30 added lines https://github.com/traefik/traefik-library-image/blob/master/.github/workflows/main.yml trafeik can prevent a production ingress outage. 

I totally agree we can't predict to the future and GitHub can turn just as evil as DockerHub but just sending a issue to the do never backlog isn't great and can loose potential customers...

**Comment by toboshii** [comment]
Just ran into this issue as well. Traefik is one of very few containers I run in my cluster that still hasn't either completely switched away from dockerhub or provides an official mirror on another registry.

**Comment by onedr0p** [comment]
For those coming across this issue I've pushed the image to a public GHCR repo.

https://github.com/k8s-at-home/container-images-mirror

It can be used in the traefik helm chart like so:

```yaml
  values:
    image:
      name: ghcr.io/k8s-at-home/traefik
```

**Comment by onedr0p** [comment]
Any update on getting traefik onboarded to the Open source developer license on Dockerhub? It seriously can't be that hard, can it? Intel, on the below repository was able to do this within 8 months. 

https://github.com/intel/intel-device-plugins-for-kubernetes/issues/633

**Comment by SantoDE** [comment]
I've personally reached out multiple times to docker already on that subject. So let's say, it's at least not fast ;-)

**Comment by onedr0p** [comment]
@emilevauge has mentioned that Traefik doesn't meet the qualifications for the Docker open source program.

https://twitter.com/emilevauge/status/1518637216632815620

With that said would it be possible to mirror Traefik to Quay or on this repo under GHCR? GHCR is free for open source projects.

**Comment by ddtmachado** [comment]
@onedr0p indeed we're investigating this possibility and the effort required to integrate with our build / CI tools

**Comment by onedr0p** [comment]
For anyone running into this issue since Traefik folks will not prioritize this annoying issue (who likes to be rate limited when rolling out changes to your ingress?), people can use mirror over at:

https://gallery.ecr.aws/docker/library/traefik

**Comment by robertpatrick** [comment]
@ddtmachado It seems like the most obvious solution that costs you nothing is to publish your images on GitHub Container Registry (i.e., ghcr.io).  Docker is attaching licensing requirements for pulling images from Docker Hub (e.g., you can use such images for commercial purposes as long as you have a valid Docker Desktop license).  I suppose we could simply fork your repo and do it ourselves but this seems like a waste for something so simple.

**Comment by ddtmachado** [comment]
@robertpatrick indeed using GHCR is simple enough technically speaking, but not future proof as they might change the policy and start charging for usage in the future.

For context I can give you the main reasons why publishing on it as an alternative to Docker Hub was not prioritized so far:

1. Publishing the official Docker image is part of our full release process, meaning we only announce a release once that image is available and that would not change.
2. Current Docker Hub rate limiting policy looks fair to us. 100 pulls per 6 hours per IP address even without a login, and it allows the free use of the images for commercial purposes for small businesses.
3. There are many workarounds that readily available and well documented, like [using a mirror](https://github.com/traefik/traefik/issues/8149#issuecomment-1176126443) or creating a local registry to cache your images.

Having said that we're still open to the possibility of adding GHCR or other alternatives, but prioritizing the time to do that over improving Traefik Proxy for the broader user base is not in the cards for the foreseeable future.

**Comment by onedr0p** [comment]
Given the recent dockerhub announcement of removing free team orgs, I'm wondering if the current org will be affected? And if so perhaps this could finally have a high priority?

https://github.com/docker/hub-feedback/issues/2314

**Comment by ldez** [comment]
The official Traefik is not affected by this because it's an official image handle by Docker it-self.

**Comment by krombel** [comment]
I think, this can be closed now as the images are now copied to ghcr.io since #10792 got implemented in #11070 or #11075.

Checkout https://github.com/traefik/traefik/pkgs/container/traefik

**Comment by nmengin** [comment]
Closed by #11070
